File: n-queens/java/NQueens.java
public class NQueens {

    private final int n;
    private long count; // Usar 'long' para evitar estouro
    private long nodes_visited; // Novo contador
    private final boolean[] cols;
    private final boolean[] leftDiags;
    private final boolean[] rightDiags;

    public NQueens(int n) {
        this.n = n;
        this.count = 0;
        this.nodes_visited = 0; // Inicializa o contador
        this.cols = new boolean[n];
        this.leftDiags = new boolean[2 * n - 1];
        this.rightDiags = new boolean[2 * n - 1];
    }

    private void solve(int row) {
        this.nodes_visited++; // Rastreia cada chamada

        if (row == n) {
            count++;
            return;
        }

        for (int col = 0; col < n; col++) {
            // Calculate diagonal indices
            int leftDiagIdx = row + col;
            int rightDiagIdx = row - col + n - 1;

            // Check if safe
            if (!cols[col] && !leftDiags[leftDiagIdx] && !rightDiags[rightDiagIdx]) {
                // Place queen
                cols[col] = true;
                leftDiags[leftDiagIdx] = true;
                rightDiags[rightDiagIdx] = true;

                // Recurse
                solve(row + 1);

                // Backtrack
                cols[col] = false;
                leftDiags[leftDiagIdx] = false;
                rightDiags[rightDiagIdx] = false;
            }
        }
    }

    public long getTotalSolutions() {
        solve(0); // Start the recursive process from row 0
        return count;
    }

    // Adiciona um getter para o contador de nós
    public long getNodesVisited() {
        return nodes_visited;
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.err.println("Usage: java NQueens <N>");
            System.exit(1);
        }

        int n = 0;
        try {
            n = Integer.parseInt(args[0]);
            if (n <= 0) {
                throw new NumberFormatException();
            }
        } catch (NumberFormatException e) {
            System.err.println("Error: <N> must be a positive integer.");
            System.exit(1);
        }

        NQueens problem = new NQueens(n);
        long total_solutions = problem.getTotalSolutions();
        long total_nodes = problem.getNodesVisited();

        // Nova saída JSON (formatada manualmente para evitar dependências)
        System.out.println(String.format(
            "{\"solutions\": %d, \"nodes_visited\": %d}",
            total_solutions,
            total_nodes
        ));
    }
}

File: n-queens/python/n_queens.py
import sys
import json

class NQueensSolver:
    def __init__(self, n):
        self.n = n
        self.count = 0
        self.nodes_visited = 0  # Novo contador
        # Helper arrays for O(1) checks
        self.cols = [False] * n
        self.left_diag = [False] * (2 * n - 1)
        self.right_diag = [False] * (2 * n - 1)

    def solve(self, row):
        """Recursive backtracking function"""
        self.nodes_visited += 1 # Rastreia cada chamada de 'solve'

        if row == self.n:
            self.count += 1
            return

        for col in range(self.n):
            # Check if the current spot is safe
            if (not self.cols[col] and
                not self.left_diag[row + col] and
                not self.right_diag[row - col + self.n - 1]):
                
                # Place queen
                self.cols[col] = True
                self.left_diag[row + col] = True
                self.right_diag[row - col + self.n - 1] = True
                
                # Recurse to the next row
                self.solve(row + 1)
                
                # Backtrack: remove queen
                self.cols[col] = False
                self.left_diag[row + col] = False
                self.right_diag[row - col + self.n - 1] = False

    def get_total_solutions(self):
        self.solve(0)
        return self.count

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 n_queens.py <N>", file=sys.stderr)
        sys.exit(1)

    try:
        n = int(sys.argv[1])
        if n <= 0:
            raise ValueError()
    except ValueError:
        print("Error: <N> must be a positive integer.", file=sys.stderr)
        sys.exit(1)

    solver = NQueensSolver(n)
    total_solutions = solver.get_total_solutions()
    
    # Nova saída JSON
    output = {
        "solutions": total_solutions,
        "nodes_visited": solver.nodes_visited
    }
    print(json.dumps(output))